xbmc/network/httprequesthandler/HTTPPythonHandler.cpp

```diff
@@ -0,0 +1,242 @@
+  {
+    m_response.type = HTTPError;
+    if (m_response.status == MHD_HTTP_FOUND)
+      m_response.status = MHD_HTTP_INTERNAL_SERVER_ERROR;
+
+    return;
+  }
+
+  m_type = boost::dynamic_pointer_cast<ADDON::CWebinterface>(m_addon)->GetType();
+
+  m_response.type = HTTPMemoryDownloadNoFreeCopy;
+  m_response.status = MHD_HTTP_OK;
+  
+  // determine the last modified date
+  const CURL pathToUrl(m_scriptPath);
+  struct __stat64 statBuffer;
+  if (XFILE::CFile::Stat(pathToUrl, &statBuffer) != 0)
+    return;
+
+  struct tm* time;
+#ifdef HAVE_LOCALTIME_R
+  struct tm result = {};
+  time = localtime_r((time_t*)&statBuffer.st_mtime, &result);
+#else
+  time = localtime((time_t *)&statBuffer.st_mtime);
+#endif
+  if (time == NULL)
+    return;
+
+  m_lastModified = *time;
+}
+
+bool CHTTPPythonHandler::CanHandleRequest(const HTTPRequest &request)
+{
+  return URIUtils::GetExtension(request.pathUrl).compare(".py") == 0;
+}
+
+int CHTTPPythonHandler::HandleRequest()
+{
+  if (m_response.type == HTTPError)
+    return MHD_YES;
+
+  std::vector<std::string> args;
+  args.push_back(m_scriptPath);
+
+  try
+  {
+    HTTPPythonRequest* pythonRequest = new HTTPPythonRequest();
+    pythonRequest->connection = m_request.connection;
+    pythonRequest->file = URIUtils::GetFileName(m_request.pathUrl);
+    CWebServer::GetRequestHeaderValues(m_request.connection, MHD_GET_ARGUMENT_KIND, pythonRequest->getValues);
+    CWebServer::GetRequestHeaderValues(m_request.connection, MHD_HEADER_KIND, pythonRequest->headerValues);
+    pythonRequest->method = m_request.method;
+    pythonRequest->postValues = m_postFields;
+    pythonRequest->requestContent = m_requestData;
+    pythonRequest->responseType = HTTPNone;
+    pythonRequest->responseLength = 0;
+    pythonRequest->responseStatus = MHD_HTTP_OK;
+    pythonRequest->url = m_request.pathUrlFull;
+    pythonRequest->path = m_request.pathUrl;
+    pythonRequest->version = m_request.version;
+    pythonRequest->requestTime = CDateTime::GetCurrentDateTime();
+    pythonRequest->lastModifiedTime = m_lastModified;
+
+    std::string hostname;
+    uint16_t port;
+    if (GetHostnameAndPort(hostname, port))
+    {
+      pythonRequest->hostname = hostname;
+      pythonRequest->port = port;
+    }
+
+    CHTTPPythonInvoker* pythonInvoker = NULL;
+    if (m_type == ADDON::WebinterfaceTypeModPython)
+      pythonInvoker = new CHTTPModPythonInvoker(&g_pythonParser, pythonRequest);
+    else
+    {
+      CLog::Log(LOGWARNING, "WebServer: unable to run python script at %s with an unknown invoker", m_scriptPath.c_str());
+
+      m_response.status = MHD_HTTP_INTERNAL_SERVER_ERROR;
+      m_response.type = HTTPError;
+
+      return MHD_YES;
+    }
+
+    LanguageInvokerPtr languageInvokerPtr(pythonInvoker);
+    int result = CScriptInvocationManager::Get().ExecuteSync(m_scriptPath, languageInvokerPtr, m_addon, args, 30000, false);
```

Comment:
I think that's right. Because iirc with Apache you can use different mod
pythons on different paths too, which essentially is the same as for us.
I.e. we hook up wsgi on a certain add-on id path and then use our image
stuff on another.